Verify some IExtensibleEnum during the connection phase

justliliandev opened the pull request:
Adds a Payload that is sent during configuration phase of a connection to verify certain extensible enums from being the same. This prevents hard to track down issues by making sure that those issues fail early with decent info for devs. The checked enums are Rarity, FireworkExplosion.Shape and GrassColorModifier. The first 2 are verified for presence/absence of values and the correct order while the last one only is verified for presence/absence of values. That's because the first 2 are serialized with the ordinal and the last by name (correct order is not needed for that one).

I currently only check those 3 as they are the only extensible enums that are sent over the network.

This is an example on how the connection error looks to the user:
![image](https://github.com/neoforged/NeoForge/assets/36055315/396569b3-0193-4634-bcc8-466e408fc282)

justliliandev committed:
Add an extensible enum checker to the configuration phase

justliliandev committed:
compute payload only once

justliliandev committed:
fix spotless

justliliandev committed:
compute payload as a lazy

justliliandev committed:
remove mobcategory verify as it is not send between client/server

KnightMiner commented:
I feel like that error needs to be more user friendly. Put the technical details in the game log, the error displayed to the user should tell them why there is an error (that is, a mismatch between server and client mods) as that will help them understand how to fix it.

It would  be nice to also print the mod IDs of the offending mods, though not sure that is tracked anywhere; perhaps names is the best option.

Shadows-of-Fire commented:
Agreed with Knight, the user-displayed error message has to contain the actionable info on how to correct the issue (by ensuring the mod list is in sync with the server).  The technical information should be sent to the log if the user has to sent it to someone else for further debugging.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/network/payload/ExtensibleEnumDataPayload.java:
```diff
@@ -0,0 +1,103 @@
+        return new ExtensibleEnumDataPayload<>(map);
+    }
+
+    private static <T extends Enum<?> & IExtensibleEnum> ExtensibleEnumDataPayload<T> factory(List<Class<? extends T>> orderedEnums, List<Class<? extends T>> unorderedEnums) {
```
Name this method `create`, since it is responsible for creation of the payload.  You'd only call a method `factory` if it created a new factory-like object that would later create something.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java:
```diff
@@ -138,4 +143,61 @@ public static void handle(AuxiliaryLightDataPayload msg, IPayloadContext context
     }
+
+    public static void handle(ExtensibleEnumDataPayload<?> msg, IPayloadContext context) {
+        enum MissMatchReason {
```
`MismatchReason`

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java:
```diff
@@ -138,4 +143,61 @@ public static void handle(AuxiliaryLightDataPayload msg, IPayloadContext context
+
+            private void buildError(StringBuilder builder, List<String> errors) {
+                String msg = switch (this) {
+                    case ID_MISMATCH -> "idmissmatch";
```
`idmismatch`

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java:
```diff
@@ -138,4 +143,61 @@ public static void handle(AuxiliaryLightDataPayload msg, IPayloadContext context
+                    case ADDED_ONLY_ON_CLIENT -> "clientOnly";
+                    case MISSING_ON_CLIENT -> "serverOnly";
+                };
+                builder.append(msg).append('[').append(String.join(",", errors)).append(']');
```
Honestly the message form translations of these errors are pretty bad.  Since this is data that should be in the log anyway, this should just reflect the long-form name of the enum constant, and should probably be joined by `", "` instead of `","` for readability.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/network/payload/ExtensibleEnumDataPayload.java:
```diff
@@ -0,0 +1,103 @@
+
+    private static ExtensibleEnumDataPayload<?> INSTANCE = null;
+
+    public static synchronized ExtensibleEnumDataPayload<?> getInstance() {
```
Rename this as `getOrCreate()`

Using a name like `getInstance` implies that there are not side effects, which isn't true for this method.

Shadows-of-Fire reviewed (changes requested):
Overall the idea is good, we can precheck this before nasty errors show up, so we certainly should.  Just need to cleanup the implementation so it makes sense in the user's context.

justliliandev committed:
improve errors, fix typos, apply requested changes

justliliandev commented:
Updated error message to the user:
![image](https://github.com/neoforged/NeoForge/assets/36055315/5aa468ed-b9a2-4341-9894-b5e6c1fb53b5)

Updated error message in the log:
```
[ne.ne.ne.ne.ha.ClientPayloadHandler$1MissMatchReason/]: Extensible Enum mismatch found: net.minecraft.world.item.Rarity has elements only on the server: myRandomRarity3
[ne.ne.ne.ne.ha.ClientPayloadHandler$1MissMatchReason/]: Extensible Enum mismatch found: net.minecraft.world.item.Rarity has elements only on the client: myRandomRarity2
[ne.ne.ne.ne.ha.ClientPayloadHandler$1MissMatchReason/]: Extensible Enum mismatch found: net.minecraft.world.item.Rarity has elements in a different order: myRandomRarity[4->5]
[minecraft/ClientCommonPacketListenerImpl]: Client disconnected with reason: Extensible Enum mismatches found. Make sure that your server and client have the same mods.
```

justliliandev commented:
> It would be nice to also print the mod IDs of the offending mods, though not sure that is tracked anywhere; perhaps names is the best option.

The origin of an extensibleenumvalue is not tracked, I would just hope that people would prefix their enum names with their modid to prevent issues. Some mods do this already,
Immersive Engineering prefix their Rarity values with `IE_`, Astral Sorcery with `AS_`, Biomancy with their full modid, Aether with `aether.`. Some mods don't do this and some use common names that could lead to duplicates and problems with that. Maybe the javadoc for IExtensibleEnum should include that people have to prefix their enum names

XFactHD commented on src/main/java/net/neoforged/neoforge/network/configuration/SyncExtensibleEnums.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public void run(Consumer<CustomPacketPayload> sender) {
+        if (listener.hasChannel(ExtensibleEnumDataPayload.TYPE)) {
```
This task must either be marked complete immediately in all cases or marked complete immediately if the client can't receive the payload and marked complete by a return payload if the client can receive this payload and did not disconnect due to any errors, otherwise the client will get stuck waiting indefinitely for this task to complete.

XFactHD commented on src/main/java/net/neoforged/neoforge/network/payload/ExtensibleEnumDataPayload.java:
```diff
@@ -0,0 +1,103 @@
+public record ExtensibleEnumDataPayload<T extends Enum<?> & IExtensibleEnum>(Map<Class<? extends T>, EnumData> enums) implements CustomPacketPayload {
+    public static final Type<ExtensibleEnumDataPayload<?>> TYPE = new Type<>(new ResourceLocation(NeoForgeVersion.MOD_ID, "extensible_enum_data"));
+    public static final StreamCodec<FriendlyByteBuf, ExtensibleEnumDataPayload<?>> STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
```
Please use a `Map<String, EnumData>` instead of two disjoint lists. The `StreamCodec` for that would be `ByteBufCodecs.map(HashMap::new, ByteBufCodecs.STRING_UTF8, EnumData.STREAM_CODEC)` and the map accessor replacing the two list accessors would be `enums.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().getName(), Map.Entry::getValue))`.

XFactHD commented on src/main/java/net/neoforged/neoforge/network/payload/ExtensibleEnumDataPayload.java:
```diff
@@ -0,0 +1,103 @@
+
+    public static synchronized ExtensibleEnumDataPayload<?> getOrCreateInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = create(List.of(Rarity.class, FireworkExplosion.Shape.class), List.of(BiomeSpecialEffects.GrassColorModifier.class));
```
`RecipeBookType` should also be synced as an ordered enum as it is used in an order-sensitive context in the `ClientboundRecipesPacket` (yes, the packet name is stupid, it doesn't actually contain any recipes, it's just recipe book info)

XFactHD reviewed (changes requested):
I agree that something like this is definitely needed, but disconnecting a client for an ID mismatch doesn't sound like a good solution to me as there is (at least to my knowledge) no guarantee that loading the game on two machines will lead to the same enum entry order, especially with parallel loading, no guidelines or guards for the timing of enum entry creation and the potential presence of one-sided mods on one of those machines, which means the end user can in the worst case not do anything about such a mismatch. Instead, something like registry sync (which can be done with the data already provided by the payload being implemented here) would be preferable in my opinion, though it would require a kind of "rebindable ID mapper" to replace the existing `BY_ID` fields in the affected enums. Using actual registries would technically work but would in my opinion be a heavy-handed and potentially error-prone solution, unless additional checks to enforce registration are added (which the only existing registry-backed extensible enum [`ItemDisplayContext`] doesn't do either).

justliliandev committed:
apply suggested improvements by xfact

justliliandev commented:
The ModClassVisitor, ModFileScanData and other places where the Entrypoint order might change are all in ordered collections. Entrypoints are executed after each other afaict. Only the DirectoryStream used to scan the mods for annotations mentions that no order is defined. So if the Enum creation for a Mod happen only in one mod entrypoint per mod, the order of the DirectoryStream doesn't matter. The second case where the order might change is in the FancyModLoader where again a DirectoryStream is used to find the modfiles, but they are sorted by name before they are processed further(see ModsFolderLocator#findCandidates)
If this is too brittle so that random smaller changes in FML might break this, then we should maybe consider fixing the order(either when the payload is sent or when new Enums are added using the RuntimeEnumExtender in FML.
The guidelines for when enum entries should be created have been made in https://github.com/neoforged/Documentation/pull/78 (create them either as a static final in your main mod entrypoint class or in your mod constructor). There are no protections to prevent this after mod construction has completed(some people have created new enum values in their Item.getRarity overrides (before the ItemStack component system, but some modders do irresponsible things)), not sure if a prevention for this should be added and it's not on me to decide something like that. For now enum values should be in the same order (if modders create them in their mod constructor) and the modfiles are not renamed by the client or the server, so that they would be sorted in a different order. 

Edit: I've been corrected, multiple mods are constructed at the same time

RaymondBlaze commented:
> If this is too brittle so that random smaller changes in FML might break this, then we should maybe consider fixing the order(either when the payload is sent or when new Enums are added using the RuntimeEnumExtender in FML.

I believe fixing the `RuntimeEnumExtender` to make it sort extended enum entries (with the internal name) would be the best practice.

> I currently only check those 3 as they are the only extensible enums that are sent over the network.

Mods might use other extensible enums in there network, so we would better be recording all enum classes that has been extended, which should also be the `RuntimeEnumExtender`'s job.

Technici4n reviewed (commented):
Why are we syncing them by ID? Shouldn't we sync them by name?

justliliandev commented:
> Why are we syncing them by ID? Shouldn't we sync them by name?

the FireworkExplosion.Shape StreamCodec in vanilly syncs by id. The FIREWORK_EXPLOSION DataComponentType uses that StreamCodec and having 2 different StreamCodecs for vanilla/modded connections for all the elements would be really annoying and fail at that DataComponentType as they are registered and only exist once and therefor their StreamCodec can only be one specific StreamCodec. StreamCodecs also have no idea for which connection they encode/decode so we can't have a wrapped StreamCodec that encodes/decodes them based on the connection. To ensure vanilla compat some StreamCodecs have to sync enum values by id

Technici4n commented:
We could sync vanilla entries by ID, and modded entries by name (using -1 as the marker ID).

justliliandev commented:
that would work for the streamcodec but not for the ClientboundRecipePacket/RecipeBookSettings which would need another solution. We would also need to know, which values were created by minecraft and which were created by mods. Which is currently not tracked and would either require some form of reflection or keeping track of vanilla/modded values by the RuntimeEnumExtender

Technici4n commented:
Superseded by the new enum extension system!

Technici4n commented:
Turns out we don't have the networking bits yet. See https://github.com/neoforged/NeoForge/issues/1133. @justliliandev Do you want to tackle that issue instead?

justliliandev commented:
yeah, I could do this probably next week
